### PR TITLE
COMMERCE-10247 - SE - fix of minicart widget

### PR DESCRIPTION
**Pull request by brianikim**
https://issues.liferay.com/browse/COMMERCE-10247

@andrea-ale-sbarra ,

cc/ @alessiorendina 

Full conversion to javascript. I also took care of all edge cases I identified.

Resent from: https://github.com/liferay-commerce/liferay-portal/pull/3137

### CODE REVIEW

**Comment by brianikim** [comment]
ci:test:relevant

**Comment by brianikim** [comment]
ci:test:sf

File: modules/apps/commerce/commerce-cart-content-web/src/main/resources/META-INF/resources/js/cart_mini/transition.js [hunk 1]
```diff
@@ -14,22 +14,20 @@
 
 export default function ({namespace}) {
 	Liferay.provide(window, `${namespace}transition`, (event) => {
-		const link = event.currentTarget;
+		const link = event.delegateTarget;
```
Comment (kresimir-coko) [inline modules/apps/commerce/commerce-cart-content-web/src/main/resources/META-INF/resources/js/cart_mini/transition.js:17]: Not 100% sure what `event` you have access to here, perhaps it's a YUI event, but if it's a [vanilla JS event](https://developer.mozilla.org/en-US/docs/Web/API/Event) then `delegateTarget` doesn't exist on it

**Review by kresimir-coko — COMMENTED** [review]
Looks fine for a quick fix, apart from that one comment about the event, but if you tested it manually then it looks fine.

**Comment by andrea-ale-sbarra** [comment]
ci:test:commerce

**Comment by andrea-ale-sbarra** [comment]
thank you @kresimir-coko

File: modules/apps/commerce/commerce-cart-content-web/src/main/resources/META-INF/resources/js/cart_mini/transition.js (hunk 1 above)
Comment (brianikim) [reply modules/apps/commerce/commerce-cart-content-web/src/main/resources/META-INF/resources/js/cart_mini/transition.js:17]: Hi @andrea-ale-sbarra  and @kresimir-coko , delegateTarget will exist always if our [view.js](modules/apps/commerce/commerce-cart-content-web/src/main/resources/META-INF/resources/js/cart_mini/view.js) is called.


See: https://github.com/liferay-commerce/liferay-portal/blob/ff484b7f5d6f46c85c6e90a2e607da3a26ad8afd/modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/delegate/delegate.es.js#L55

@andrea-ale-sbarra , let me know if everything is good.

Comment (kresimir-coko) [reply modules/apps/commerce/commerce-cart-content-web/src/main/resources/META-INF/resources/js/cart_mini/transition.js:17]: IDK the intricacies of your infrastructure, but you linked to a `delegate` utility which needs to be imported and used as specified in the API:
```js
/**
 * Listens to the specified event on the given DOM element, but only calls the
 * given callback listener when it's triggered by elements that match the
 * given selector or target element.
 * @param {!Element} element The DOM element the event should be listened on.
 * @param {string} eventName The name of the event to listen to.
 * @param {string} selector Css selector that should match the event for the
 *     listener to be triggered.
 * @param {!function(!Object)} callback Function to be called when the event
 *     is triggered. It will receive the normalized event object.
 * @return {function} Can be used to remove the listener.
 */
function delegate(element, eventName, selector, callback) {
```

It does not magically enable `event.delegateTarget` to be used wherever in the codebase.

This needs to be manually tested by whoever is in charge.

Comment (brianikim) [reply modules/apps/commerce/commerce-cart-content-web/src/main/resources/META-INF/resources/js/cart_mini/transition.js:17]: @kresimir-coko , if you look at commit: https://github.com/liferay-commerce/liferay-portal/pull/3161/commits/e247c8c0fcd52b343e12b70436995b3f47b25067 and see the full file, you will already see delegate is being imported and used as specified in the API.

Comment (brianikim) [reply modules/apps/commerce/commerce-cart-content-web/src/main/resources/META-INF/resources/js/cart_mini/transition.js:17]: Also, I have already manually tested this.

Comment (kresimir-coko) [reply modules/apps/commerce/commerce-cart-content-web/src/main/resources/META-INF/resources/js/cart_mini/transition.js:17]: 👍

File: modules/apps/commerce/commerce-cart-content-web/src/main/resources/META-INF/resources/js/cart_mini/transition.js [hunk 2]
```diff
@@ -41,12 +39,13 @@ export default function ({namespace}) {
 			`${namespace}transitionComments`
 		);
 
+		transitionComments.classList.remove('hide');
+
 		transitionComments.style.display = 'block';
 
 		const dialog = Liferay.Util.Window.getWindow({
```
Comment (brianikim) [inline modules/apps/commerce/commerce-cart-content-web/src/main/resources/META-INF/resources/js/cart_mini/transition.js:46]: @kresimir-coko , it looks like I am no longer able to use Liferay.Util.Window.getWindow as a function (no longer exists?). I get following error:`Uncaught TypeError: Liferay.Util.Window.getWindow is not a function`. I am requesting your assistance in refactoring the dialog here as front-end is not my strongest point. Could you let me know best way to proceed?
`

Comment (brianikim) [reply modules/apps/commerce/commerce-cart-content-web/src/main/resources/META-INF/resources/js/cart_mini/transition.js:46]: Hello @kresimir-coko , I believe I did most of the conversion correctly. Please correct me if the following needs improvement. I have one question for how to replace `bodyContent: form` correctly?
```
		openModal
			({
				buttons: [
				{
					displayType: 'primary',
					label: Liferay.Language.get('done'),
					onClick() {
						{
							submitForm(form);
						}
					},
				},
				{
					autoFocus: true,
					displayType: 'secondary',
					label: Liferay.Language.get('cancel'),
					type: 'cancel',
				},
			],
			height: '40vh',
			id: `${namespace}transitionFm`,
			size: 'lg',
			title: link.innerText
		});
	});

```

**Comment by brianikim** [comment]
Reassigning the ticket to myself as more changes are needed on the javascript end.

File: modules/apps/commerce/commerce-cart-content-web/src/main/resources/META-INF/resources/js/cart_mini/transition.js (hunk 2 above)
Comment (kresimir-coko) [reply modules/apps/commerce/commerce-cart-content-web/src/main/resources/META-INF/resources/js/cart_mini/transition.js:46]: > it looks like I can no longer use `Liferay.Util.Window.GetWindow` 

There’s 25 usages of this in the codebase, it wasn’t removed, so it should work like it did before.

[This test demonstrates](https://github.com/liferay/liferay-portal/blob/master/modules/apps/frontend-js/frontend-js-web/test/liferay/modal/Modal.js#L106-L110) that if you are passing it a `bodyComponent` prop, it needs to be a React component. 

Other option is to use the `bodyHTML` prop and pass HTML to it, maybe you could try using that instead of `bodyContent: form` and replace it with `bodyHTML: form`.

I am not an expert here, and I haven't migrated that many of these to know how it all works without testing it in the code myself. But, it still looks to me that you shouldn't be getting the `Uncaught TypeError: Liferay.Util.Window.getWindow is not a function` error and you might have a different problem with your code.

**Comment by brianikim** [comment]
ci:test:relevant

**Comment by brianikim** [comment]
ci:test:sf

**Comment by brianikim** [comment]
@kresimir-coko , see latest commit.

File: modules/apps/commerce/commerce-cart-content-web/package.json [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+	"dependencies": {
+		"frontend-js-web": "*"
+	},
+	"name": "@liferay/commerce-cart-content-web",
+	"scripts": {
+		"build": "liferay-npm-scripts build",
+		"checkFormat": "liferay-npm-scripts check",
+		"format": "liferay-npm-scripts fix"
+	},
+	"version": "4.0.38"
```
Comment (andrea-ale-sbarra) [inline modules/apps/commerce/commerce-cart-content-web/package.json:11]: @brianikim @kresimir-coko still I don't get why this is not something like 1.0.0

Comment (kresimir-coko) [reply modules/apps/commerce/commerce-cart-content-web/package.json:11]: If in doubt, ask in a public channel. I've got no definitive answer here.

File: modules/apps/commerce/commerce-cart-content-web/src/main/resources/META-INF/resources/js/cart_mini/transition.js [hunk 4]
```diff
@@ -13,87 +13,93 @@
...
+			const workflowTaskId = parseInt(link.dataset.workflowtaskid, 10);
 
-		const form = document.getElementById(`${namespace}transitionFm`);
+			const form = document.getElementById(`${namespace}transitionFm`);
 
-		document.getElementById(`${namespace}transitionCommerceOrderId`).value =
-			link.dataset.commerceorderid;
+			document.getElementById(
+				`${namespace}transitionCommerceOrderId`
+			).value = link.dataset.commerceorderid;
```
Comment (kresimir-coko) [inline modules/apps/commerce/commerce-cart-content-web/src/main/resources/META-INF/resources/js/cart_mini/transition.js:28]: Extract the doc.getelbyid into its own variable and check if it exists before assigning it the link.dataset.commerceorderid.

File: modules/apps/commerce/commerce-cart-content-web/src/main/resources/META-INF/resources/js/cart_mini/transition.js [hunk 5]
```diff
@@ -13,87 +13,93 @@
...
-		document.getElementById(
-			`${namespace}workflowTaskId`
-		).value = workflowTaskId;
-		document.getElementById(`${namespace}transitionName`).value =
-			link.dataset.transitionname;
+			document.getElementById(
+				`${namespace}workflowTaskId`
+			).value = workflowTaskId;
+			document.getElementById(`${namespace}transitionName`).value =
+				link.dataset.transitionname;
```
Comment (kresimir-coko) [inline modules/apps/commerce/commerce-cart-content-web/src/main/resources/META-INF/resources/js/cart_mini/transition.js:34]: Same as above, extract in a variable and check if the variable exists. It's called a sanity check. It's defensive programming to make sure JS doesn't break due to a silly error.

**Review by kresimir-coko — CHANGES_REQUESTED** [review]
Looks good apart from the couple minor things. Glad you found a way to make it work :)

File: modules/apps/commerce/commerce-cart-content-web/src/main/resources/META-INF/resources/js/cart_mini/transition.js (hunk 5 above)
Comment (brianikim) [reply modules/apps/commerce/commerce-cart-content-web/src/main/resources/META-INF/resources/js/cart_mini/transition.js]: Done as requested.

File: modules/apps/commerce/commerce-cart-content-web/src/main/resources/META-INF/resources/js/cart_mini/transition.js (hunk 4 above)
Comment (brianikim) [reply modules/apps/commerce/commerce-cart-content-web/src/main/resources/META-INF/resources/js/cart_mini/transition.js]: Done as requested.

**Review by kresimir-coko — APPROVED** [review]

**Comment by pietrobocale** [comment]
Resent to #3219

**Comment by andrea-ale-sbarra** [comment]
Will be forwarded with this PR https://github.com/liferay-commerce/liferay-portal/pull/3219
